quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/SanitizableConfigSource.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.configuration;
+
+import org.keycloak.quarkus.runtime.configuration.mappers.PropertyMappers;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Config source with the ability to sanitize its properties based on environment setup
+ */
+public interface SanitizableConfigSource {
```

Comment:
> The required state of the app is that after disabling particular props, we should consider them as such as they have never existed.

This is so we can skip the validation of those values, and also not persist them - without introducing new calls into the respective logic to check for options being disabled correct?

> With the filtering, they should still exist somewhere, and if we call directly the config source, they'd be present there.

That's currently needed for the Picocli.validateConfig in some form - with the sanitize approach via KeycloakConfigSourceProvider.getSanitizedRemovedProperties(). There'd be something similar on the interceptor, or the interceptor would be temporarily disabled like PropertyMappingInterceptor.

Does your concern about the interceptor stem from other cases where you see config consumers by-passing top-level config calls and going directly against individual config sources?

> @shawkins I'm kinda on top of a fence and not sure, what's the most suitable approach. I'd probably prefer to keep the current approach as we can save some time on it. And we can do some amendments in follow-up tasks. WDYT?

Let's loop in @radcortez as well.

Reply:
> This is so we can skip the validation of those values, and also not persist them - without introducing new calls into the respective logic to check for options being disabled correct?

This is a little bit tricky. We need to distinguish between 2 phases of the initialization:
1. **BEFORE** the evaluation of disabled mappers - we need to have properly initialized all config sources as normal; to be able to determine, whether the mapper will be disabled or not. 
2. **AFTER** the evaluation of disabled mappers - return values of only relevant mappers

To explain my intention more - in the first prototype, I put the check of disabled mappers in relevant places. However, I've decided not to use it as such places might be a lot and it could potentially cause some undesirable behavior. So I decided to sanitize our config sources for good after the evaluation of disabled mappers.

Did it answer your question, or you meant sth else?

Reply:
> Does your concern about the interceptor stem from other cases where you see config consumers by-passing top-level config calls and going directly against individual config sources?

If you need to omit a particular property an interceptor is the best alternative. Interceptors are executed before sources are queried and are decoupled from sources, so you can add or remove behavior without changing the sources themselves.

For the interceptor you only need to filter the property being queried in `getValue` and `iterateNames`. Now, if someone gets a reference to the specific source, they could still retrieve the filtered names, but usually, this is not something that users should do.

Reply:
@radcortez Thanks for your input. 

Just a few questions:
 - So, we don't need to override the `iterateValues`, is it correct?
 -  If the interceptor in extensions does not have any `@Priority` annotation, what is the default priority?
 - Is it `Priorities.APPLICATION=5000` ? (and potentially dynamically handled conflicts on the SmallRye side?)

Reply:
> * So, we don't need to override the `iterateValues`, is it correct?

No. We don't use it, it is deprecated and we will remove it soon.

> * If the interceptor in extensions does not have any `@Priority` annotation, what is the default priority?
Its `Priorities#APPLICATION=5000`

> * Is it `Priorities.APPLICATION=5000` ? (and potentially dynamically handled conflicts on the SmallRye side?)
Yes. If multiple interceptors are found with the same priority, the behavior is nondeterministic. All the interceptors are added to the chain, but we don't know the execution order (for the ones sharing the priority)

Reply:
@radcortez Ok, thanks